src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs

```diff
@@ -140,53 +142,56 @@ public partial class NetworkChange
         private static unsafe void CreateSocket()
         {
             Debug.Assert(s_socket == null, "s_socket is not null, must close existing socket before opening another.");
-            int newSocket;
+            IntPtr newSocket;
             Interop.Error result = Interop.Sys.CreateNetworkChangeListenerSocket(&newSocket);
             if (result != Interop.Error.SUCCESS)
             {
                 string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
                 throw new NetworkInformationException(message);
             }
 
-            s_socket = new SocketWrapper(newSocket);
+            s_socket = new Socket(new SafeSocketHandle(newSocket, ownsHandle: true));
 
-            new Thread(args => LoopReadSocket((SocketWrapper)args!))
-            {
-                IsBackground = true,
-                Name = ".NET Network Address Change"
-            }.UnsafeStart(s_socket);
+            // Don't capture ExecutionContext.
+            ThreadPool.UnsafeQueueUserWorkItem(static socket =>
+                _ = ReadEventsAsync(socket),
+                s_socket, preferLocal: false);
         }
 
         private static void CloseSocket()
         {
             Debug.Assert(s_socket != null, "s_socket was null when CloseSocket was called.");
-            Interop.Error result = Interop.Sys.CloseNetworkChangeListenerSocket(s_socket != null ? s_socket.Socket : -1);
-            if (result != Interop.Error.SUCCESS)
-            {
-                string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
-                throw new NetworkInformationException(message);
-            }
-
+            s_socket.Dispose();
             s_socket = null;
         }
 
-        private static unsafe void LoopReadSocket(SocketWrapper initiallyCreatedSocket)
+        private static async Task ReadEventsAsync(Socket socket)
         {
-            while (s_socket == initiallyCreatedSocket) //if both references are equal then s_socket was not changed
+            while (s_socket == socket)
             {
-                //we can continue processing events
-                Interop.Sys.ReadEvents(initiallyCreatedSocket.Socket, &ProcessEvent);
+                // Wait for data to become available.
+                await socket.ReceiveAsync(Array.Empty<byte>(), SocketFlags.None).ConfigureAwait(false);
+
+                if (!TryReadEvents(socket))
+                {
+                    return; // Stop reading on error/EOF.
+                }
             }
+
+            static unsafe bool TryReadEvents(Socket socket)
+                => Interop.Sys.ReadEvents(socket.SafeHandle, &ProcessEvent) >= 0;
```

Comment:
Seems like there are currently race conditions here, e.g. we could be about to call ReadEvents, then the socket / SafeHandle are disposed, and this will then throw ObjectDisposedException.

Reply:
Yes, depending on when the socket gets disposed, either `ObjectDisposedException` gets thrown or `SocketException` with `SocketError.OperationAborted`.